arch/x86/core/intel64/locore.S

```diff
@@ -365,6 +365,23 @@ __resume:
 	movq $0xB9, _thread_offset_to_rip(%rdi)
 #endif
 
+#ifdef CONFIG_THREAD_LOCAL_STORAGE
+	/*
+	 * Write the TLS base pointer to FS_BASE MSR,
+	 * where GCC emits code to access TLS data via
+	 * offset to FS.
+	 * Since wrmsr write EDX:EAX to MSR indicated by
+	 * ECX, the high 32-bit needs to be loaded into
+	 * RDX and right shifted by 32 bits so EDX has
+	 * the higher 32-bit value.
+	 */
+	movl $X86_FS_BASE, %ecx
+	movq _thread_offset_to_tls(%rdi), %rax
+	movq _thread_offset_to_tls(%rdi), %rdx
+	shrq $32, %rdx
+	wrmsr
```

Comment:
Could use wrfsbase instruction on targets with support for that?

Reply:
QEMU raises invalid opcode with `wrfsbase` even if `CR4.FSGSBASE` is enabled. So this can't be used at the moment.

Reply:
interesting. my experience with QEMU's x86 emulation has been near-perfect, so perhaps this is a true limitation